### PR TITLE
Prefer importlib.metadata over pkg_resources if available

### PR DESCRIPTION
**Pull request by sentrivana**
Let's not use `pkg_resources` and avoid a `DeprecationWarning` if `importlib.metadata` is available.

Fixes https://github.com/getsentry/sentry-python/issues/2048

### CODE REVIEW

File: sentry_sdk/integrations/modules.py [hunk 1]
```diff
@@ -21,12 +21,20 @@
 def _generate_installed_modules():
     # type: () -> Iterator[Tuple[str, str]]
     try:
-        import pkg_resources
-    except ImportError:
-        return
+        from importlib.metadata import distributions, version
+
+        for dist in distributions():
+            yield dist.metadata["Name"], version(dist.metadata["Name"])
```
Comment (sentrivana) [inline sentry_sdk/integrations/modules.py:27]: There's also `dist.name` (it's just an alias of `dist.metadata["Name"]`), but it hasn't always been there. `dist.metadata["Name"]` should work with all py versions.

**Comment by sentrivana** [comment]
> If `distributions()` yields the same stuff as `pkg_resources.working_set` this looks good and can be merged.

@antonpirker It does with one difference: `distributions()` keeps the capitalization, `working_set` makes everything lowercase, e.g.:
```
# distributions()
MarkupSafe 2.1.2
matplotlib-inline 0.1.6
mock 5.0.2
packaging 23.1
parso 0.8.3
pexpect 4.8.0
pickleshare 0.7.5
pip 23.1.2
platformdirs 3.5.0
pluggy 0.13.1
prompt-toolkit 3.0.38
ptyprocess 0.7.0
pure-eval 0.2.2
py 1.11.0
Pygments 2.15.1
```

```
# working_set
markupsafe 2.1.2
matplotlib-inline 0.1.6
mock 5.0.2
packaging 23.1
parso 0.8.3
pexpect 4.8.0
pickleshare 0.7.5
pip 23.1.2
platformdirs 3.5.0
pluggy 0.13.1
prompt-toolkit 3.0.38
ptyprocess 0.7.0
pure-eval 0.2.2
py 1.11.0
pygments 2.15.1
```

Do you see this creating issues anywhere? I grepped for where we actually use the modules and it seems like we [attach them to the event](https://github.com/getsentry/sentry-python/blob/master/sentry_sdk/integrations/modules.py#L55) and [check them here](https://github.com/getsentry/sentry-python/blob/master/sentry_sdk/integrations/asgi.py#L111) but elsewhere we seem to be using `sys.modules` instead.

**Review by antonpirker — APPROVED** [review]

**Comment by antonpirker** [comment]
@sentrivana could you please add a test for testing `_generate_installed_modules`?

File: sentry_sdk/integrations/modules.py [hunk 2]
```diff
@@ -18,15 +18,29 @@
 _installed_modules = None
 
 
+def _normalize_module_name(name):
```
Comment (antoni-szych-rtbhouse) [inline sentry_sdk/integrations/modules.py:21]:
```suggestion
def _normalize_module_name(name: str):
```

Comment (antonpirker) [reply sentry_sdk/integrations/modules.py:21]: we can not have types this way, because we still support Python 2.7.

**Comment by sentrivana** [comment]
@antonpirker Changed the PR a bit and added a test, could you please skim over to see if it still looks ok to you?

File: sentry_sdk/integrations/modules.py [hunk 3]
```diff
@@ -18,15 +18,30 @@
...
+    # type: (str) -> str
+    return name.lower().replace("-", "_")
```
[the remark this answers is not in the document]
Comment (sentrivana) [reply sentry_sdk/integrations/modules.py:23]: There's a weird mismatch between `importlib` and `pkg_resources` where they report the package `typing-extensions` differently; `importlib` reports `typing_extensions`, while `pkg_resources` reports it as `typing-extensions`. This is the only package that is different (if we normalize to lowercase, which we're doing). I'll remove the `.replace()` and change the test to ignore the one package.

[the remark this answers is not in the document]
Comment (antonpirker) [reply sentry_sdk/integrations/modules.py]: I guess if we just try to keep as much the same as it was reported with `pkg_resources` we are good to go.

**Pull request merged**